File: java-dsa/src/main/java/io/giovannymassuia/trie/Trie.java
package io.giovannymassuia.trie;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.Supplier;

public class Trie {
    private final TrieNode root = new TrieNode();

    public void insert(String path, Supplier<Boolean> handler) {
        TrieNode current = root;
        String[] segments = path.split("/");
        Set<String> seenParameters = new HashSet<>();

        for (String segment : segments) {
            if (segment.startsWith(":")) {
                String paramName = segment.substring(1);

                if (seenParameters.contains(paramName)) {
                    throw new IllegalArgumentException("Cannot use the same parameter name more than once in a route.");
                }
                seenParameters.add(paramName);

                if (current.parameterChild != null) {
                    // Check if the parameter is of the same type.
                    if (!current.parameterChild.parameterName.equals(paramName)) {
                        throw new IllegalArgumentException("Invalid path: " + path + ". Parameter " + paramName + " is already defined.");
                    }
                } else {
                    TrieNode paramNode = new TrieNode();
                    paramNode.parameterName = paramName;
                    current.parameterChild = paramNode;
                }
                current = current.parameterChild;
            } else {
                current.staticChildren.putIfAbsent(segment, new TrieNode());
                current = current.staticChildren.get(segment);
            }
        }
        current.handler = handler;
    }

    public SearchResult search(String path) {
        TrieNode current = root;
        String[] segments = path.split("/");
        Map<String, String> capturedParams = new HashMap<>();

        for (String segment : segments) {
            TrieNode next = current.staticChildren.get(segment);

            // If exact match is not found, check if a parameter child exists.
            if (next == null && current.parameterChild != null) {
                next = current.parameterChild;
                capturedParams.put(next.parameterName, segment);
            }

            if (next == null) {
                return new SearchResult(() -> false, null);
            }

            current = next;
        }
        return new SearchResult(current.handler, capturedParams);
    }

    public record SearchResult(Supplier<Boolean> handler, Map<String, String> params) {
    }

}

File: go-dsa/trie/trie.go
package trie

import (
	"errors"
	"strings"
)

type TrieNode struct {
	StaticChildren map[string]*TrieNode
	ParameterChild *TrieNode
	Handler        func() bool
	ParameterName  string // Changed from pointer to string
	HasParam       bool   // New field to indicate if ParameterName is set
}

type Trie struct {
	Root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{
		Root: newTrieNode(),
	}
}

func newTrieNode() *TrieNode {
	return &TrieNode{
		StaticChildren: make(map[string]*TrieNode),
	}
}

func (t *Trie) Insert(path string, handler func() bool) error {
	current := t.Root
	segments := splitPath(path)
	seenParameters := make(map[string]bool)

	for _, segment := range segments {
		if segment[0] == ':' {
			paramName := segment[1:]

			if seenParameters[paramName] {
				return errors.New("cannot use the same parameter name more than once in a route")
			}
			seenParameters[paramName] = true

			if current.ParameterChild != nil {
				if current.ParameterChild.ParameterName != paramName {
					return errors.New("Invalid path: " + path + ". Parameter " + paramName + " is already defined.")
				}
			} else {
				paramNode := newTrieNode()
				paramNode.ParameterName = paramName
				paramNode.HasParam = true
				current.ParameterChild = paramNode
			}

			current = current.ParameterChild
		} else {
			if _, exists := current.StaticChildren[segment]; !exists {
				current.StaticChildren[segment] = newTrieNode()
			}

			current = current.StaticChildren[segment]
		}
	}

	current.Handler = handler

	return nil
}

type SearchResult struct {
	Handler func() bool
	Params  map[string]string
}

func (t *Trie) Search(path string) (result SearchResult) {
	current := t.Root
	segments := splitPath(path)
	capturedParams := make(map[string]string)

	for _, segment := range segments {
		next, exists := current.StaticChildren[segment]

		if !exists && current.ParameterChild != nil && current.ParameterChild.HasParam {
			next = current.ParameterChild
			capturedParams[current.ParameterChild.ParameterName] = segment
		}

		if next == nil {
			return SearchResult{Handler: func() bool { return false }, Params: nil}
		}

		current = next
	}

	return SearchResult{Handler: current.Handler, Params: capturedParams}
}

func splitPath(path string) []string {
	return strings.Split(path, "/")[1:]
}
